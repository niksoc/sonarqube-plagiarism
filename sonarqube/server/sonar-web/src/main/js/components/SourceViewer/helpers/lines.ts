/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { intersection } from 'lodash';

export function optimizeHighlightedSymbols(
  symbolsForLine: string[] = [],
  highlightedSymbols: string[] = []
): string[] | undefined {
  const symbols = intersection(symbolsForLine, highlightedSymbols);

  return symbols.length ? symbols : undefined;
}

export function optimizeLocationMessage(
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined,
  optimizedSecondaryIssueLocations: T.LinearIssueLocation[]
) {
  return highlightedLocationMessage != null &&
    optimizedSecondaryIssueLocations.some(
      location => location.index === highlightedLocationMessage.index
    )
    ? highlightedLocationMessage
    : undefined;
}

export function optimizeSelectedIssue(selectedIssue: string | undefined, issuesForLine: T.Issue[]) {
  return selectedIssue !== undefined && issuesForLine.find(issue => issue.key === selectedIssue)
    ? selectedIssue
    : undefined;
}
