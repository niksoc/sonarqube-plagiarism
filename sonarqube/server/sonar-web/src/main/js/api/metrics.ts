/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export interface MetricsResponse {
  metrics: T.Metric[];
  p: number;
  ps: number;
  total: number;
}

export function getMetrics(data?: {
  isCustom?: boolean;
  p?: number;
  ps?: number;
}): Promise<MetricsResponse> {
  return getJSON('/api/metrics/search', data).catch(throwGlobalError);
}

export function getAllMetrics(data?: {
  isCustom?: boolean;
  p?: number;
  ps?: number;
}): Promise<T.Metric[]> {
  return inner(data);

  function inner(
    data: { p?: number; ps?: number } = { ps: 500 },
    prev?: MetricsResponse
  ): Promise<T.Metric[]> {
    return getMetrics(data).then(r => {
      const result = prev ? prev.metrics.concat(r.metrics) : r.metrics;
      if (r.p * r.ps >= r.total) {
        return result;
      }
      return inner({ ...data, p: r.p + 1 }, { ...r, metrics: result });
    });
  }
}

export function getMetricDomains(): Promise<string[]> {
  return getJSON('/api/metrics/domains').then(r => r.domains, throwGlobalError);
}

export function getMetricTypes(): Promise<string[]> {
  return getJSON('/api/metrics/types').then(r => r.types, throwGlobalError);
}

export function createMetric(data: {
  description?: string;
  domain?: string;
  key: string;
  name: string;
  type: string;
}): Promise<T.Metric> {
  return postJSON('/api/metrics/create', data).catch(throwGlobalError);
}

export function updateMetric(data: {
  description?: string;
  domain?: string;
  id: string;
  key?: string;
  name?: string;
  type?: string;
}) {
  return post('/api/metrics/update', data).catch(throwGlobalError);
}

export function deleteMetric(data: { keys: string }) {
  return post('/api/metrics/delete', data).catch(throwGlobalError);
}
