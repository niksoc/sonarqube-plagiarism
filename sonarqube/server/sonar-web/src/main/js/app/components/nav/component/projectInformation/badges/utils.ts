/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { stringify } from 'querystring';
import { getLocalizedMetricName } from 'sonar-ui-common/helpers/l10n';
import { omitNil } from 'sonar-ui-common/helpers/request';
import { getHostUrl, getPathUrlAsString } from 'sonar-ui-common/helpers/urls';
import { getProjectUrl } from '../../../../../../helpers/urls';

export type BadgeColors = 'white' | 'black' | 'orange';
export type BadgeFormats = 'md' | 'url';

export interface BadgeOptions {
  branch?: string;
  color?: BadgeColors;
  format?: BadgeFormats;
  project?: string;
  metric?: string;
  pullRequest?: string;
}

export enum BadgeType {
  measure = 'measure',
  qualityGate = 'quality_gate',
  marketing = 'marketing'
}

export function getBadgeSnippet(type: BadgeType, options: BadgeOptions) {
  const url = getBadgeUrl(type, options);
  const { branch, format = 'md', metric = 'alert_status', project } = options;

  if (format === 'url') {
    return url;
  } else {
    let label;
    let projectUrl;

    switch (type) {
      case BadgeType.marketing:
        label = 'SonarCloud';
        break;
      case BadgeType.measure:
        label = getLocalizedMetricName({ key: metric });
        break;
      case BadgeType.qualityGate:
      default:
        label = 'Quality gate';
        break;
    }

    if (project) {
      projectUrl = getPathUrlAsString(getProjectUrl(project, branch), false);
    }

    const mdImage = `![${label}](${url})`;
    return projectUrl ? `[${mdImage}](${projectUrl})` : mdImage;
  }
}

export function getBadgeUrl(
  type: BadgeType,
  { branch, project, color = 'white', metric = 'alert_status', pullRequest }: BadgeOptions
) {
  switch (type) {
    case BadgeType.marketing:
      return `${getHostUrl()}/images/project_badges/sonarcloud-${color}.svg`;
    case BadgeType.qualityGate:
      return `${getHostUrl()}/api/project_badges/quality_gate?${stringify(
        omitNil({ branch, project, pullRequest })
      )}`;
    case BadgeType.measure:
    default:
      return `${getHostUrl()}/api/project_badges/measure?${stringify(
        omitNil({ branch, project, metric, pullRequest })
      )}`;
  }
}
