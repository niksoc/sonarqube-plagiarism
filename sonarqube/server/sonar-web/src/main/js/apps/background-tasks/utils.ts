/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { toShortNotSoISOString } from 'sonar-ui-common/helpers/dates';
import { ALL_TYPES, CURRENTS, STATUSES } from './constants';

export interface Query {
  currents: string;
  maxExecutedAt?: Date;
  minSubmittedAt?: Date;
  query: string;
  status: string;
  taskType: string;
}

export function updateTask(tasks: T.Task[], newTask: T.Task) {
  return tasks.map(task => (task.id === newTask.id ? newTask : task));
}

export function mapFiltersToParameters(filters: Partial<Query> = {}) {
  const parameters: any = {};

  if (filters.status === STATUSES.ALL) {
    parameters.status = [
      STATUSES.PENDING,
      STATUSES.IN_PROGRESS,
      STATUSES.SUCCESS,
      STATUSES.FAILED,
      STATUSES.CANCELED
    ].join();
  } else if (filters.status === STATUSES.ALL_EXCEPT_PENDING) {
    parameters.status = [
      STATUSES.IN_PROGRESS,
      STATUSES.SUCCESS,
      STATUSES.FAILED,
      STATUSES.CANCELED
    ].join();
  } else {
    parameters.status = filters.status;
  }

  if (filters.taskType !== ALL_TYPES) {
    parameters.type = filters.taskType;
  }

  if (filters.currents !== CURRENTS.ALL) {
    parameters.onlyCurrents = true;
  }

  if (filters.minSubmittedAt) {
    parameters.minSubmittedAt = toShortNotSoISOString(filters.minSubmittedAt);
  }

  if (filters.maxExecutedAt) {
    parameters.maxExecutedAt = toShortNotSoISOString(filters.maxExecutedAt);
  }

  if (filters.query) {
    parameters.q = filters.query;
  }

  return parameters;
}

const ONE_SECOND = 1000;
const ONE_MINUTE = 60 * ONE_SECOND;
const ONE_HOUR = 60 * ONE_MINUTE;

function format(int: number, suffix: string) {
  return `${int}${suffix}`;
}

export function formatDuration(value: number | undefined) {
  if (!value) {
    return '';
  }
  if (value < ONE_SECOND) {
    return format(value, 'ms');
  } else if (value < ONE_SECOND * 10) {
    const seconds = Math.floor(value / ONE_SECOND);
    const ms = value - seconds * ONE_SECOND;
    return seconds + '.' + format(ms, 's');
  } else if (value < ONE_MINUTE) {
    const seconds = Math.floor(value / ONE_SECOND);
    return format(seconds, 's');
  } else if (value < ONE_MINUTE * 10) {
    const minutes = Math.floor(value / ONE_MINUTE);
    const seconds = Math.floor((value - minutes * ONE_MINUTE) / ONE_SECOND);
    return format(minutes, 'min') + ' ' + format(seconds, 's');
  }
  const hours = Math.floor(value / ONE_HOUR);
  const minutes = Math.floor((value - hours * ONE_HOUR) / ONE_MINUTE);
  return format(hours, 'h') + ' ' + format(minutes, 'min');
}
