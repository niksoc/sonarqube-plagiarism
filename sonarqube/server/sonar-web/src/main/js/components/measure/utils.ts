/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getRatingTooltip as nextGetRatingTooltip, isDiffMetric } from '../../helpers/measures';
import { getLeakPeriod } from '../../helpers/periods';

const KNOWN_RATINGS = [
  'sqale_rating',
  'maintainability_rating', // Needed to provide the label for "new_maintainability_rating"
  'reliability_rating',
  'security_rating'
];

export function enhanceMeasure(measure: T.Measure, metrics: T.Dict<T.Metric>): T.MeasureEnhanced {
  return {
    ...measure,
    metric: metrics[measure.metric],
    leak: getLeakValue(measure)
  };
}

export function getLeakValue(measure: T.MeasureIntern | undefined): string | undefined {
  if (!measure || !measure.periods) {
    return undefined;
  }
  const period = getLeakPeriod(measure.periods);
  return period && period.value;
}

export function getRatingTooltip(metricKey: string, value: number): string | undefined {
  const finalMetricKey = isDiffMetric(metricKey) ? metricKey.substr(4) : metricKey;
  if (KNOWN_RATINGS.includes(finalMetricKey)) {
    return nextGetRatingTooltip(finalMetricKey, value);
  }
  return undefined;
}
