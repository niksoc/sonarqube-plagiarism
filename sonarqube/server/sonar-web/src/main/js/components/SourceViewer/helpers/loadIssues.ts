/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { searchIssues } from '../../../api/issues';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { parseIssueFromResponse } from '../../../helpers/issues';
import { BranchLike } from '../../../types/branch-like';

// maximum possible value
const PAGE_SIZE = 500;

function buildQuery(component: string, branchLike: BranchLike | undefined) {
  return {
    additionalFields: '_all',
    resolved: 'false',
    componentKeys: component,
    s: 'FILE_LINE',
    ...getBranchLikeQuery(branchLike)
  };
}

export function loadPage(
  query: T.RawQuery,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return searchIssues({
    ...query,
    p: page,
    ps: pageSize
  }).then(r =>
    r.issues.map(issue => parseIssueFromResponse(issue, r.components, r.users, r.rules))
  );
}

export function loadPageAndNext(
  query: T.RawQuery,
  toLine: number,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return loadPage(query, page).then(issues => {
    if (issues.length === 0) {
      return [];
    }

    const lastIssue = issues[issues.length - 1];

    if (
      (lastIssue.textRange != null && lastIssue.textRange.endLine > toLine) ||
      issues.length < pageSize
    ) {
      return issues;
    }

    return loadPageAndNext(query, toLine, page + 1, pageSize).then(nextIssues => {
      return [...issues, ...nextIssues];
    });
  });
}

export default function loadIssues(
  component: string,
  _fromLine: number,
  toLine: number,
  branchLike: BranchLike | undefined
): Promise<T.Issue[]> {
  const query = buildQuery(component, branchLike);
  return new Promise(resolve => {
    loadPageAndNext(query, toLine, 1).then(issues => {
      resolve(issues);
    });
  });
}
