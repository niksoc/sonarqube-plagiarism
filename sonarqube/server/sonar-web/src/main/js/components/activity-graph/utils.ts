/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { chunk, flatMap, groupBy, sortBy } from 'lodash';
import { getLocalizedMetricName, translate } from 'sonar-ui-common/helpers/l10n';
import { get, save } from 'sonar-ui-common/helpers/storage';
import { localizeMetric } from '../../helpers/measures';
import { MetricKey } from '../../types/metrics';
import { GraphType, MeasureHistory, Serie } from '../../types/project-activity';

export const DEFAULT_GRAPH = GraphType.issues;

const GRAPHS_METRICS_DISPLAYED: T.Dict<string[]> = {
  [GraphType.issues]: [MetricKey.bugs, MetricKey.code_smells, MetricKey.vulnerabilities],
  [GraphType.coverage]: [MetricKey.lines_to_cover, MetricKey.uncovered_lines],
  [GraphType.duplications]: [MetricKey.ncloc, MetricKey.duplicated_lines]
};

const GRAPHS_METRICS: T.Dict<string[]> = {
  [GraphType.issues]: GRAPHS_METRICS_DISPLAYED[GraphType.issues].concat([
    MetricKey.reliability_rating,
    MetricKey.security_rating,
    MetricKey.sqale_rating
  ]),
  [GraphType.coverage]: [...GRAPHS_METRICS_DISPLAYED[GraphType.coverage], MetricKey.coverage],
  [GraphType.duplications]: [
    ...GRAPHS_METRICS_DISPLAYED[GraphType.duplications],
    MetricKey.duplicated_lines_density
  ]
};

export function isCustomGraph(graph: GraphType) {
  return graph === GraphType.custom;
}

export function getGraphTypes(ignoreCustom = false) {
  const graphs = [GraphType.issues, GraphType.coverage, GraphType.duplications];
  return ignoreCustom ? graphs : [...graphs, GraphType.custom];
}

export function hasDataValues(serie: Serie) {
  return serie.data.some(point => Boolean(point.y || point.y === 0));
}

export function hasHistoryData(series: Serie[]) {
  return series.some(serie => serie.data && serie.data.length > 1);
}

export function getSeriesMetricType(series: Serie[]) {
  return series.length > 0 ? series[0].type : 'INT';
}

export function getDisplayedHistoryMetrics(graph: GraphType, customMetrics: string[]) {
  return isCustomGraph(graph) ? customMetrics : GRAPHS_METRICS_DISPLAYED[graph];
}

export function getHistoryMetrics(graph: GraphType, customMetrics: string[]) {
  return isCustomGraph(graph) ? customMetrics : GRAPHS_METRICS[graph];
}

export function hasHistoryDataValue(series: Serie[]) {
  return series.some(serie => serie.data && serie.data.length > 1 && hasDataValues(serie));
}

export function splitSeriesInGraphs(series: Serie[], maxGraph: number, maxSeries: number) {
  return flatMap(
    groupBy(series, serie => serie.type),
    type => chunk(type, maxSeries)
  ).slice(0, maxGraph);
}

export function generateCoveredLinesMetric(
  uncoveredLines: MeasureHistory,
  measuresHistory: MeasureHistory[]
) {
  const linesToCover = measuresHistory.find(measure => measure.metric === MetricKey.lines_to_cover);
  return {
    data: linesToCover
      ? uncoveredLines.history.map((analysis, idx) => ({
          x: analysis.date,
          y: Number(linesToCover.history[idx].value) - Number(analysis.value)
        }))
      : [],
    name: 'covered_lines',
    translatedName: translate('project_activity.custom_metric.covered_lines'),
    type: 'INT'
  };
}

export function generateSeries(
  measuresHistory: MeasureHistory[],
  graph: GraphType,
  metrics: T.Metric[] | T.Dict<T.Metric>,
  displayedMetrics: string[]
): Serie[] {
  if (displayedMetrics.length <= 0 || measuresHistory === undefined) {
    return [];
  }
  return sortBy(
    measuresHistory
      .filter(measure => displayedMetrics.indexOf(measure.metric) >= 0)
      .map(measure => {
        if (measure.metric === MetricKey.uncovered_lines && !isCustomGraph(graph)) {
          return generateCoveredLinesMetric(measure, measuresHistory);
        }
        const metric = findMetric(measure.metric, metrics);
        return {
          data: measure.history.map(analysis => ({
            x: analysis.date,
            y: metric && metric.type === 'LEVEL' ? analysis.value : Number(analysis.value)
          })),
          name: measure.metric,
          translatedName: metric ? getLocalizedMetricName(metric) : localizeMetric(measure.metric),
          type: metric ? metric.type : 'INT'
        };
      }),
    serie =>
      displayedMetrics.indexOf(serie.name === 'covered_lines' ? 'uncovered_lines' : serie.name)
  );
}

export function saveActivityGraph(
  namespace: string,
  project: string,
  graph: GraphType,
  metrics: string[] = []
) {
  save(namespace, graph, project);
  if (isCustomGraph(graph)) {
    save(`${namespace}.custom`, metrics.join(','), project);
  }
}

export function getActivityGraph(
  namespace: string,
  project: string
): { graph: GraphType; customGraphs: string[] } {
  const customGraphs = get(`${namespace}.custom`, project);
  return {
    graph: (get(namespace, project) as GraphType) || DEFAULT_GRAPH,
    customGraphs: customGraphs ? customGraphs.split(',') : []
  };
}

function findMetric(key: string, metrics: T.Metric[] | T.Dict<T.Metric>) {
  if (Array.isArray(metrics)) {
    return metrics.find(metric => metric.key === key);
  }
  return metrics[key];
}
