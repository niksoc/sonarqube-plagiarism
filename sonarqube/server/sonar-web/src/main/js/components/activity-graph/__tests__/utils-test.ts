/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable sonarjs/no-duplicate-string */
import * as dates from 'sonar-ui-common/helpers/dates';
import { MetricKey } from '../../../types/metrics';
import { GraphType, Serie } from '../../../types/project-activity';
import * as utils from '../utils';

jest.mock('date-fns/start_of_day', () =>
  jest.fn(date => {
    const startDay = new Date(date);
    startDay.setUTCHours(0, 0, 0, 0);
    return startDay;
  })
);

const HISTORY = [
  {
    metric: MetricKey.lines_to_cover,
    history: [
      { date: dates.parseDate('2017-04-27T08:21:32.000Z'), value: '100' },
      { date: dates.parseDate('2017-04-30T23:06:24.000Z'), value: '100' }
    ]
  },
  {
    metric: MetricKey.uncovered_lines,
    history: [
      { date: dates.parseDate('2017-04-27T08:21:32.000Z'), value: '12' },
      { date: dates.parseDate('2017-04-30T23:06:24.000Z'), value: '50' }
    ]
  }
];

const METRICS = [
  { id: '1', key: MetricKey.uncovered_lines, name: 'Uncovered Lines', type: 'INT' },
  { id: '2', key: MetricKey.lines_to_cover, name: 'Line to Cover', type: 'PERCENT' }
];

const SERIE: Serie = {
  data: [
    { x: dates.parseDate('2017-04-27T08:21:32.000Z'), y: 2 },
    { x: dates.parseDate('2017-04-28T08:21:32.000Z'), y: 2 }
  ],
  name: 'foo',
  translatedName: 'Foo',
  type: 'PERCENT'
};

describe('generateCoveredLinesMetric', () => {
  it('should correctly generate covered lines metric', () => {
    expect(utils.generateCoveredLinesMetric(HISTORY[1], HISTORY)).toMatchSnapshot('with data');
    expect(utils.generateCoveredLinesMetric(HISTORY[1], [])).toMatchSnapshot('empty data');
  });
});

describe('generateSeries', () => {
  it('should correctly generate the series', () => {
    expect(
      utils.generateSeries(HISTORY, GraphType.coverage, METRICS, [
        MetricKey.uncovered_lines,
        MetricKey.lines_to_cover
      ])
    ).toMatchSnapshot();
  });
  it('should correctly handle non-existent data', () => {
    expect(utils.generateSeries(HISTORY, GraphType.coverage, METRICS, [])).toEqual([]);
  });
});

describe('getDisplayedHistoryMetrics', () => {
  const customMetrics = ['foo', 'bar'];
  it('should return only displayed metrics on the graph', () => {
    expect(utils.getDisplayedHistoryMetrics(utils.DEFAULT_GRAPH, [])).toEqual([
      MetricKey.bugs,
      MetricKey.code_smells,
      MetricKey.vulnerabilities
    ]);
    expect(utils.getDisplayedHistoryMetrics(GraphType.coverage, customMetrics)).toEqual([
      MetricKey.lines_to_cover,
      MetricKey.uncovered_lines
    ]);
  });
  it('should return all custom metrics for the custom graph', () => {
    expect(utils.getDisplayedHistoryMetrics(GraphType.custom, customMetrics)).toEqual(
      customMetrics
    );
  });
});

describe('getHistoryMetrics', () => {
  const customMetrics = ['foo', 'bar'];
  it('should return all metrics', () => {
    expect(utils.getHistoryMetrics(utils.DEFAULT_GRAPH, [])).toEqual([
      MetricKey.bugs,
      MetricKey.code_smells,
      MetricKey.vulnerabilities,
      MetricKey.reliability_rating,
      MetricKey.security_rating,
      MetricKey.sqale_rating
    ]);
    expect(utils.getHistoryMetrics(GraphType.coverage, customMetrics)).toEqual([
      MetricKey.lines_to_cover,
      MetricKey.uncovered_lines,
      GraphType.coverage
    ]);
    expect(utils.getHistoryMetrics(GraphType.custom, customMetrics)).toEqual(customMetrics);
  });
});

describe('hasHistoryData', () => {
  it('should correctly detect if there is history data', () => {
    expect(
      utils.hasHistoryData([
        {
          name: 'foo',
          translatedName: 'foo',
          type: 'INT',
          data: [
            { x: dates.parseDate('2017-04-27T08:21:32.000Z'), y: 2 },
            { x: dates.parseDate('2017-04-30T23:06:24.000Z'), y: 2 }
          ]
        }
      ])
    ).toBeTruthy();
    expect(
      utils.hasHistoryData([
        {
          name: 'foo',
          translatedName: 'foo',
          type: 'INT',
          data: []
        },
        {
          name: 'bar',
          translatedName: 'bar',
          type: 'INT',
          data: [
            { x: dates.parseDate('2017-04-27T08:21:32.000Z'), y: 2 },
            { x: dates.parseDate('2017-04-30T23:06:24.000Z'), y: 2 }
          ]
        }
      ])
    ).toBeTruthy();
    expect(
      utils.hasHistoryData([
        {
          name: 'bar',
          translatedName: 'bar',
          type: 'INT',
          data: [{ x: dates.parseDate('2017-04-27T08:21:32.000Z'), y: 2 }]
        }
      ])
    ).toBeFalsy();
  });
});

describe('getGraphTypes', () => {
  it('should correctly return the graph types', () => {
    expect(utils.getGraphTypes()).toMatchSnapshot();
    expect(utils.getGraphTypes(true)).toMatchSnapshot();
  });
});

describe('hasDataValues', () => {
  it('should check for data value', () => {
    expect(utils.hasDataValues(SERIE)).toBe(true);
    expect(utils.hasDataValues({ ...SERIE, data: [] })).toBe(false);
  });
});

describe('getSeriesMetricType', () => {
  it('should return the correct type', () => {
    expect(utils.getSeriesMetricType([SERIE])).toBe('PERCENT');
    expect(utils.getSeriesMetricType([])).toBe('INT');
  });
});

describe('hasHistoryDataValue', () => {
  it('should return the correct type', () => {
    expect(utils.hasHistoryDataValue([SERIE])).toBe(true);
    expect(utils.hasHistoryDataValue([])).toBe(false);
  });
});
