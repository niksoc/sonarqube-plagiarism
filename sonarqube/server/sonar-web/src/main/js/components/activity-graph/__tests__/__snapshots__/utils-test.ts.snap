// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCoveredLinesMetric should correctly generate covered lines metric: empty data 1`] = `
Object {
  "data": Array [],
  "name": "covered_lines",
  "translatedName": "project_activity.custom_metric.covered_lines",
  "type": "INT",
}
`;

exports[`generateCoveredLinesMetric should correctly generate covered lines metric: with data 1`] = `
Object {
  "data": Array [
    Object {
      "x": 2017-04-27T08:21:32.000Z,
      "y": 88,
    },
    Object {
      "x": 2017-04-30T23:06:24.000Z,
      "y": 50,
    },
  ],
  "name": "covered_lines",
  "translatedName": "project_activity.custom_metric.covered_lines",
  "type": "INT",
}
`;

exports[`generateSeries should correctly generate the series 1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "x": 2017-04-27T08:21:32.000Z,
        "y": 88,
      },
      Object {
        "x": 2017-04-30T23:06:24.000Z,
        "y": 50,
      },
    ],
    "name": "covered_lines",
    "translatedName": "project_activity.custom_metric.covered_lines",
    "type": "INT",
  },
  Object {
    "data": Array [
      Object {
        "x": 2017-04-27T08:21:32.000Z,
        "y": 100,
      },
      Object {
        "x": 2017-04-30T23:06:24.000Z,
        "y": 100,
      },
    ],
    "name": "lines_to_cover",
    "translatedName": "Line to Cover",
    "type": "PERCENT",
  },
]
`;

exports[`getGraphTypes should correctly return the graph types 1`] = `
Array [
  "issues",
  "coverage",
  "duplications",
  "custom",
]
`;

exports[`getGraphTypes should correctly return the graph types 2`] = `
Array [
  "issues",
  "coverage",
  "duplications",
]
`;
